README.md
---

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test Coverage][coveralls-image]][coveralls-url]
[![Dependency Status][dep-image]][dep-url]
[![devDependency Status][devdep-image]][devdep-url] 
[![NPM downloads][downloads-image]][npm-url]

[![Sauce Test Status][sauce-image]][sauce-url]

[npm-image]: http://img.shields.io/npm/v/uxcore-totop.svg?style=flat-square
[npm-url]: http://npmjs.org/package/uxcore-totop
[travis-image]: https://img.shields.io/travis/uxcore/uxcore-totop.svg?style=flat-square
[travis-url]: https://travis-ci.org/uxcore/uxcore-totop
[coveralls-image]: https://img.shields.io/coveralls/uxcore/uxcore-totop.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/uxcore/uxcore-totop?branch=master
[dep-image]: http://img.shields.io/david/uxcore/uxcore-totop.svg?style=flat-square
[dep-url]: https://david-dm.org/uxcore/uxcore-totop
[devdep-image]: http://img.shields.io/david/dev/uxcore/uxcore-totop.svg?style=flat-square
[devdep-url]: https://david-dm.org/uxcore/uxcore-totop#info=devDependencies
[downloads-image]: https://img.shields.io/npm/dm/uxcore-totop.svg
[sauce-image]: https://saucelabs.com/browser-matrix/uxcore-totop.svg
[sauce-url]: https://saucelabs.com/u/uxcore-totop


#### setup develop environment

```sh
$ git clone https://github.com/uxcore/uxcore-totop
$ cd uxcore-totop
$ npm install
$ gulp server
```

## Usage

```javascript
<Totop>
    <div className="block1">了解我们</div>
    <div className="block2"><i className="kuma-icon kuma-icon-comment"></i></div>
</Totop>
```

## demo
http://uxcore.github.io/

## API

## Props

| 配置项 | 类型 | 必填 | 默认值 | 功能/备注 |
|---|---|---|---|---|
|prefixCls|string|optional|"kuma-totop"|不想使用 kuma 主题时使用|
|className|string|optional|-|额外的顶级类名|
|to|number|optional|10|回到顶部时据顶端距离，单位 px|
|distance|number|optional|30|向下滑多少距离出现回到顶部|
|duration|number|optional|600|动画持续时间|

## TotopBox props

| 配置项 | 类型 | 必填 | 默认值 | 功能/备注 |
|---|---|---|---|---|
|label|string|require|||
|icon|React.Element|require||推荐使用iconfont,见demo|
|onClick|function|optional|||


HISTORY.md
# 0.6.1

* `CHANGED` support js style export

# 0.6.0

* `CHANGED` adapt react@16.x]

# 0.5.0

* `CHANGED` update react version 15.x

# 0.4.0

* `CHANGED` new icon

# 0.3.0

* `NEW` add new prop onTotopEnd
* `NEW` add tests
* `CHANGED` update `uxcore-popover` to `~0.4.0`
* `CHANGED` use `unstable_renderSubtreeIntoContainer`

# 0.2.0

`CHANGED` new style

# 0.1.2

`CHANGED` update scaffold

# 0.1.1

`FIXED` fix ie8 scrollTop bug 


tests/Totop.spec.js
import expect from 'expect.js';
import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import sinon from 'sinon';
import $ from 'jquery';
import Totop from '../src';

sinon.spy(Totop.prototype, 'handleGotopClick');

const getWrapperElement = () => {
    var div = document.createElement('div');
    document.body.appendChild(div);
    return div;
};

function createTotopAndScroll(to, duration, callback) {
    let totop;
    ReactDOM.render(
        <div style={{height: 3000}}>
            <Totop ref={(c) => {totop = c}}/>
        </div>,
        getWrapperElement()
    );
    let el = document.body;
    el.scrollTop = 2000;
    totop.scrollTo(el, to, duration, () => {
        callback && callback(el, totop);
    });
}

describe('create Totop', () => {
    let wrapper;

    it('should render gotop button', () => {
        wrapper = mount(
            <Totop />
        );
        expect(wrapper.find('.gotop-box')).to.have.length(1);
    });

    it('should trigger Totop click event', () => {
        const wrapper = mount(
            <Totop />
        );
        wrapper.find('.gotop-box .btn').simulate('click');
        expect(Totop.prototype.handleGotopClick.calledOnce).to.be(true);
    });

    it('should scroll to correct position', (done) => {
        createTotopAndScroll(10, 200, () => {
            expect(document.body.scrollTop).to.be(10);
            done();
        });
    });

    it('should show Totop when body scrollTop larger than distance', (done) => {
        const distance = 100;
        createTotopAndScroll(200, 200, (el, instance) => {
            expect(el.scrollTop > distance).to.be(true);
            done();
        });
    });

    /*it('should get correct scrollTop when timer', () => {
        wrapper = mount(
            <Totop distance={100} />
        );
        const instance = wrapper.instance();
        document.body.scrollTop = 1000;
        setTimeout(() => {

        }, 10);
    });*/

    it('should remove scroll.totop event when componentWillUnmount', () => {
        wrapper = mount(
            <Totop distance={100} />
        );

        const instance = wrapper.instance();
        sinon.spy(Totop.prototype, 'componentWillUnmount');
        wrapper.unmount();
        expect(Totop.prototype.componentWillUnmount.calledOnce).to.be(true);
    });
});

tests/TotopBox.spec.js
import expect from 'expect.js';
import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import sinon from 'sinon';
import TotopBox from '../src/TotopBox';

describe('create TotopBox', () => {
    let wrapper;

    it('should be able to show view without any options', () => {
        wrapper = mount(
            <TotopBox />
        );
        expect(wrapper.find('.box-text')).to.have.length(1);
    });

    it('should render icon that with classname', () => {
        const Icon = <svg className="svg-icon" />;
        wrapper = mount(
            <TotopBox icon={Icon} />
        );
        expect(wrapper.find('.box-icon.svg-icon').length).to.be(1);
    });

    it('should render label', () => {
        const label = 'label';
        wrapper = mount(
            <TotopBox label={label} />
        );
        expect(wrapper.find('.box-text').text()).to.be(label);
    });

    it('simulate click events', () => {
        const clickHandler = sinon.spy();
        wrapper = mount(
            <TotopBox onClick={clickHandler} />
        );
        wrapper.find('.box').simulate('click');
        expect(clickHandler.calledOnce).to.equal(true);
    });
});

src/TotopBox.js
const classnames = require('classnames');
const React = require('react');

class TotopBox extends React.Component {

  render() {
    let me = this;
    let iconClass = 'box-icon';
    let iconPropClass = me.props.icon && me.props.icon.props && me.props.icon.props.className;
    if (iconPropClass) {
      iconClass = classnames('box-icon', {[iconPropClass]: !!iconPropClass});
    }
    return (<div className={classnames({
      'box': true,
      [me.props.className]: !!me.props.className
    })} onClick={me.props.onClick.bind(me)}>
      <div className="box-window">
        <span className="box-text">{me.props.label}</span>
        {
          me.props.icon && React.cloneElement(me.props.icon, {
            className: iconClass
          })
        }
      </div>
    </div>)
  }
}

TotopBox.displayName = 'TotopBox';

TotopBox.propTypes = {
  className: React.PropTypes.string,
  label: React.PropTypes.string,
  icon: React.PropTypes.element,
  onClick: React.PropTypes.func
};

TotopBox.defaultProps = {
  className: '',
  label: '',
  icon: null,
  onClick: () => {}
};

module.exports = TotopBox;


demo/TotopDemo.js
/**
 * Totop Component Demo for uxcore
 * @author eternaslky
 *
 * Copyright 2014-2015, Uxcore Team, Alinw.
 * All rights reserved.
 */

let classnames = require('classnames');

let Totop = require('../src');
let {Box} = Totop;
let Popover = require('uxcore-popover');

function getRandomColor() {
  var letters = '0123456789ABCDEF'.split('');
  var color = '#';
  for (var i = 0; i < 6; i++) {
    color += letters[Math.floor(Math.random() * 16)];
  }
  return color;
}

let len = Math.random() * 10 + 10;
let mockData = [];
for (let i = 0; i < len; i++) {
  mockData.push({
    index: i,
    color: getRandomColor()
  })
}

class Demo extends React.Component {

  constructor(props) {
    super(props);
    this.state = {}
  }

  onBoxClick(box) {
    console.log(box + ' clicked!');
  }

  render() {
    // console.log(mockData);
    let overlay = <div>
      <div className="demoContent">
        <i className="kuma-icon kuma-icon-information"></i>
        <span>这是一个气泡弹窗</span>
      </div>
    </div>;
    return (
      <div>
        <Totop>
          <Box label="评价" icon={<i className="kuma-icon kuma-icon-shoucang1"></i>} onClick={this.onBoxClick.bind(this, 'pingjia')}/>
          <Popover title="这是标题，可以隐藏" overlay={overlay} placement="left">
            <div className="box">
              <div className="box-window">
                <span className="box-text">咨询</span>
                <i className="kuma-icon kuma-icon-query box-icon"></i>
              </div>
            </div>
          </Popover>
        </Totop>
        {mockData.map((item, index) => {
          return <div key={index} className="demoDiv" style={{background: item.color}}>{"第" + item.index + "个区块"}</div>
        })}
      </div>
    );
  }
}

module.exports = Demo;


src/Totop.js
/**
 * Totop Component for uxcore
 * @author eternaslky
 *
 * Copyright 2014-2015, Uxcore Team, Alinw.
 * All rights reserved.
 */

const classnames = require('classnames');
const React = require('react');
const ReactDOM = require('react-dom');
const Box = require('./TotopBox');

class Totop extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            showTotop: false
        }
    }

    componentDidMount() {
        let me = this;
        $(window).on("scroll.totop", function(e) {
            var y = (window.pageYOffset !== undefined) ? window.pageYOffset : (document.documentElement || document.body.parentNode || document.body).scrollTop;
            if (y > me.props.distance && !me.state.showTotop) {
                me.setState({
                    showTotop: true
                })
            }
            else if (y <= me.props.distance && me.state.showTotop) {
                me.setState({
                    showTotop: false
                })
            }
        })
    }

    componentWillUnmount() {
        let me = this;
        clearTimeout(me.timer);
        $(window).off("scroll.totop");
    }

    /*
     * scroll method to action like jQuery animation.
     * @param element {DOM} scroll element
     * @param to {number} the final scrollTop you want
     * @param duration {number} scroll animation time (ms)
     */

    scrollTo(element, to, duration) {
        let me = this;
        if (duration <= 0) return;
        var difference = to - element.scrollTop;
        var perTick = difference / duration * 10;

        me.timer = setTimeout(function() {
            element.scrollTop = element.scrollTop + perTick;
            if (element.scrollTop === to) return;
            me.scrollTo(element, to, duration - 10);
        }, 10);
    }

    handleGotopClick() {
        let me = this;
        me.scrollTo((document.body.scrollTop != 0 ? document.body : document.documentElement), me.props.to, me.props.duration);
    }

    render() {
        let me = this;
        let showCls = me.props.prefixCls
        return (
            <div className={classnames({
                [me.props.prefixCls]: true,
                [me.props.className]: !!me.props.className,
                [me.props.theme]: !!me.props.theme,
                "fn-clear": true
            })}>
              <div className={classnames({
                    "box gotop-box": true,
                    "show": me.state.showTotop
                })}>
                <a className="box-window btn" onClick={me.handleGotopClick.bind(me)}>
                  <span className="box-text">顶部</span>
                  <i className="kuma-icon kuma-icon-jiantou-copy box-icon"></i>
                </a>
              </div>
              {me.props.children}
            </div>
        );
    }

}

Totop.defaultProps = {
    prefixCls: 'kuma-totop',
    to: 10,
    duration: 600,
    distance: 30
};


// http://facebook.github.io/react/docs/reusable-components.html
Totop.propTypes = {
    prefixCls: React.PropTypes.string,
    className: React.PropTypes.string,
    to: React.PropTypes.number,
    distance: React.PropTypes.number,
    duration: React.PropTypes.number
};

Totop.displayName = "Totop";

Totop.Box = Box;

module.exports = Totop;
